UIAutomation: Adjusted setUp() and setUpWithError() methods to speed up execution

pachlava opened the pull request:
UI Tests execution sped up by 110 seconds (~466 seconds vs ~570 seconds)

### Fix
1. Adjusted `setUp()` and `setUpWithError()` methods to avoid app launch before every test
2. For this created method that attempts returning to 'base' screen for every test: `getToAllNotes()` (for Note Editor tests)

Also (not related to speed):
3. Created several methods like `AllNotesAssert.notesExist() `on top of `AllNotesAssert.noteExists()` to allow passing an array of note names instead of calling the method N times
4. Added guard statements to several old methods

### Test
1. Running all tests and seeing no new fails introduced, with execution time being within 460 sec range.

### Review
Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

pachlava committed:
Execution sped up by 110 seconds.

- Adjusted setUp() and setUpWithError() to avoid app launch before every test
- For this created method that attempts returning to 'base' screen for every test: getToAllNotes() (for Note Editor tests)
- Created several methods like AllNotes.createNotes() on top of existing AllNotes.createNoteAndLeaveEditor()

pachlava committed:
Fixed a typo and added a constant

^ This

mokagio commented on Simplenote.xcodeproj/xcshareddata/xcschemes/SimplenoteUITests.xcscheme:
```diff
@@ -21,6 +21,77 @@
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testCanFlipToEditMode()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testCanPreviewMarkdownBySwiping()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testCreateCheckedItem()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testCreateUncheckedItem()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testLongTappingOnLinkOpensLinkInNewWindow()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testTappingOnLinkInPreviewOpensLinkInNewWindow()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testUndoUndoesTheLastEdit()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsNoteEditor/testUsingInsertChecklistInsertsChecklist()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsTrash/testCanDeleteNoteForeverIndividually()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsTrash/testCanDeleteNotesForeverViaEmpty()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsTrash/testCanRestoreNote()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsTrash/testCanTrashNote()">
+               </Test>
+               <Test
+                  Identifier = "SimplenoteUISmokeTestsTrash/testCanViewTrashedNotes()">
+               </Test>
+            </SelectedTests>
```
Is this list of selected tests intentional?

pachlava replied:
This change appeared as soon as I opened the scheme and viewed the 'Tests' section. I did not change anything in the selection of tests. I guess it was initialized this way.

mokagio commented on SimplenoteUITests/AllNotes.swift:
```diff
@@ -5,29 +5,35 @@ class AllNotes {
         app.tables.cells[noteName].tap()
     }
-    
+
```
🎉 for removing trailing whitespaces. 👍 

Totally useless for how the code work, but sticklers like me really appreciate 🤣 🙇‍♂️

mokagio commented on SimplenoteUITests/AllNotes.swift:
```diff
@@ -5,29 +5,35 @@ class AllNotes {
         NoteEditor.leaveEditor()
     }
 
+    class func createNotes(noteNamesArray: Array<String>) {
```
Nitpick. No need to address this, just wanted to point out that, while `Array<String>` is correct, `[String]` is the most common way of writing this.

`[String]` is syntax sugar for `Array<String>`.

pachlava replied:
Thanks! Will update this during next set of changes :-)

mokagio commented on SimplenoteUITests/AllNotes.swift:
```diff
@@ -5,29 +5,35 @@ class AllNotes {
         NoteEditor.leaveEditor()
     }
 
+    class func createNotes(noteNamesArray: Array<String>) {
```
Another nitpick.

The [Swift API Guidelines](https://swift.org/documentation/api-design-guidelines/#weak-type-information) recommend to precede each weakly typed parameter, such as `Any`, `Int`, or `String, with a noun describing its role:

<img width="500" alt="screen shot 2019-02-06 at 1 43 58 pm" src="https://user-images.githubusercontent.com/1218433/52317755-78bf6600-2a15-11e9-93ba-ec4a22fb71cb.png">

In this case, `Array<String>` is clear enough that appending "array" to the name is note necessary.

What do you think of this name?

```suggestion
    class func createNotes(names: Array<String>) {
```

pachlava replied:
Thanks! Applied

mokagio commented on SimplenoteUITests/UIDs.swift:
```diff
@@ -2,6 +2,7 @@
 let uidPicture_AppLogo = "simplenote-logo"
 
 let uidNavBar_AllNotes = "All Notes"
+let uidNavBar_LogIn = "Log In"
```
I should have noticed this in the previous review but I didn't, so sorry for the out of scope suggestion. No need to address it now, just a nice-to-have tidiness idea for later.

Swift uses camelcase, there are no instances that I know of where we use `_` to separate parts of the name of a value or method. The only exception is that some testing practitioners do use `_` to separate test components in the test method name, but that's an exception and is not even a standard practice.

Anyways... What I'm trying to say is that `uidNavBar_LogIn` is not "swifty". One option could be to simply remove it and use `uidNavBarLogIn`. Another suggestion, more interesting in my opinion, would be to leverage Swift nested types feature and do this:

```swift
enum UID {

    enum NavBar {
        static let allNotes = "All Notes"
        static let logIn = "Log In"
        // ...
    }

    enum Buttons {
        static let allNotes = "All Notes"
        static let trash = "Trash"
        // ...
    }
}
```

Using an `enum` without any case means that it won't be able to instantiated (you can't do `_ = UID()`).

An extra benefit is that Xcode will help with its autocompletion:

![2021-02-22 20-29-14 2021-02-22 20_31_25](https://user-images.githubusercontent.com/1218433/108689567-57bf0480-754d-11eb-9181-ab223a2bce9d.gif)

pachlava replied:
Thanks for bringing this up! I realized this will be discussed sooner or later. Changed this to nested type as you suggested. The UIDs file now looks much tidier 🎉

mokagio reviewed (approved):
Pending input on the change in the scheme, this is good to go 👍 

```
Executed 23 tests, with 0 failures (0 unexpected) in 443.029 (480.719) seconds
```

<img alt="image" src="https://user-images.githubusercontent.com/1218433/108689738-8dfc8400-754d-11eb-9be3-40816191596b.png" width=300/>

pachlava committed:
Update SimplenoteUITests/AllNotes.swift

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Now UIDs for locators are nested types

- Changed UIDs structure from a constant: uidTextField_Password to nested types: UID.TextField.Password
- Replaced Array<String> with [String]

pachlava commented:
@mokagio thanks! I applied everything you have proposed. Actually every comment of yours made the code shorter, this I something I really like!
